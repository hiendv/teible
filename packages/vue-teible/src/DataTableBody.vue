<template>
  <tbody>
    <tr
      v-for="(d, index) in items" :key="index" :class="{
        [theme.datatable__row]: true,
        [theme['datatable__row--odd']]: index % 2 === 1
      }"
      data-elm="row" :data-odd="index % 2 === 1" @click="click($event, d, index)"
      @mouseenter="hover($event, d, index)" @mouseleave="hover($event, d, index, true)"
    >
      <data-table-cell
        v-for="(column, columnIndex) in columns" :key="columnIndex" v-bind="column.attrs"
        :item="d" :column="column" :index="index"
        :class="{
          [theme.datatable__cell]: true,
          [theme['datatable__cell--last-column']]: columnIndex === columns.length - 1,
          [theme['datatable__cell--last-row']]: index === items.length - 1
        }"
        data-elm="cell" :data-last-column="columnIndex === columns.length - 1" :data-last-row="index === items.length - 1"
      />
    </tr>
  </tbody>
</template>
<script>
import DataTableCell from './DataTableCell'

export default {
  name: 'DataTableBody',
  components: { DataTableCell },
  props: {
    items: {
      type: Array,
      required: true
    },
    columns: {
      type: Array,
      required: true
    },
    click: {
      type: Function,
      required: true
    },
    hover: {
      type: Function,
      required: true
    }
  },
  inject: ['$theme'],
  computed: {
    theme () {
      return this.$theme()
    }
  }
}
</script>
