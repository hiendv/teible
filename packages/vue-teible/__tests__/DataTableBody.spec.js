import { shallowMount } from '@vue/test-utils'
import idObj from 'identity-obj-proxy'
import DataTableBody from '../src/DataTableBody.vue'

describe('DataTableBody', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(DataTableBody, {
      propsData: {
        items: [{
          one: 'foo',
          two: 'bar'
        }, {
          one: 'baz',
          two: 'qux'
        }],
        columns: [{
          field: 'one',
          label: 'One'
        }, {
          field: 'two',
          label: 'Two'
        }],
        click: () => {},
        hover: () => {}
      },
      provide: () => ({
        $theme: () => idObj
      })
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('accepts click', () => {
    const click = jest.fn()

    const wrapper = shallowMount(DataTableBody, {
      propsData: {
        items: [{
          one: 'foo',
          two: 'bar'
        }, {
          one: 'baz',
          two: 'qux'
        }],
        columns: [{
          field: 'one',
          label: 'One'
        }, {
          field: 'two',
          label: 'Two'
        }],
        click,
        hover: () => {}
      },
      provide: () => ({
        $theme: () => idObj
      })
    })

    wrapper.find('tr').trigger('click')
    return wrapper.vm.$nextTick()
      .then(() => {
        expect(click.mock.calls.length).toBe(1)
        expect(click.mock.calls[0][0]).toBeInstanceOf(MouseEvent)
        expect(click.mock.calls[0][0].type).toBe('click')
        expect(click.mock.calls[0][1]).toStrictEqual({ one: 'foo', two: 'bar' })
        expect(click.mock.calls[0][2]).toStrictEqual(0)
      })
  })

  it('accepts hover', () => {
    const hover = jest.fn()

    const wrapper = shallowMount(DataTableBody, {
      propsData: {
        items: [{
          one: 'foo',
          two: 'bar'
        }, {
          one: 'baz',
          two: 'qux'
        }],
        columns: [{
          field: 'one',
          label: 'One'
        }, {
          field: 'two',
          label: 'Two'
        }],
        click: () => {},
        hover
      },
      provide: () => ({
        $theme: () => idObj
      })
    })

    wrapper.find('tr').trigger('hover')
    return wrapper.vm.$nextTick()
      .then(() => {
        expect(hover.mock.calls.length).toBe(1)
        expect(hover.mock.calls[0][0]).toBeInstanceOf(Event)
        expect(hover.mock.calls[0][0].type).toBe('hover')
        expect(hover.mock.calls[0][1]).toStrictEqual({ one: 'foo', two: 'bar' })
        expect(hover.mock.calls[0][2]).toStrictEqual(0)
      })
  })
})
