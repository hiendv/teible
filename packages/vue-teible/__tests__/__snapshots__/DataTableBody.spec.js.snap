// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableBody renders correctly 1`] = `
<tbody>
  <tr data-elm="row" class="datatable__row">
    <data-table-cell-stub item="[object Object]" column="[object Object]" data-elm="cell" class="[object Object]"></data-table-cell-stub>
    <data-table-cell-stub item="[object Object]" column="[object Object]" data-elm="cell" data-last-column="true" class="[object Object]"></data-table-cell-stub>
  </tr>
  <tr data-elm="row" data-odd="true" class="datatable__row datatable__row--odd">
    <data-table-cell-stub item="[object Object]" column="[object Object]" data-elm="cell" data-last-row="true" class="[object Object]"></data-table-cell-stub>
    <data-table-cell-stub item="[object Object]" column="[object Object]" data-elm="cell" data-last-column="true" data-last-row="true" class="[object Object]"></data-table-cell-stub>
  </tr>
</tbody>
`;
